### PR TITLE
Add mechanism to turn any RAII guard into a Python Context Manager

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #102037

This PR:
- adds a mechanism to turn any RAII guard into a Python Context Manager
- turns ExcludeDispatchKeyGuard into a context manager, and purges usages
of the older torch._C.ExcludeDispatchKeyGuard from the codebase.

The mechanism is that given a RAII guard, we construct a context
manager object that holds an optional guard. When we enter the context
manager we populate the guard, when we exit we reset it.

We don't delete torch._C.ExcludeDispatchKeyGuard for BC reasons (people
are using it in fbcode). If this code actually sticks
(it is using C++17 and that worries me a bit), then I'll apply the
change to other RAII guards we have, otherwise, we can write our own
std::apply.

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

File: torch/csrc/utils/python_raii.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+template <typename GuardT, typename... Args>
+struct RAIIContextManager {
+  explicit RAIIContextManager(Args&&... args)
+      : args_(std::forward<Args>(args)...) {}
+
+  void enter() {
+    auto emplace = [&](Args... args) {
+      return guard_.emplace(std::forward<Args>(args)...);
+    };
+    std::apply(emplace, args_);
```
Comment (Skylion007) [inline torch/csrc/utils/python_raii.h:17]:
```suggestion
    std::apply(std::move(emplace), args_);
```

Comment (zou3519) [reply torch/csrc/utils/python_raii.h:17]: Does this require the captures to be movable as well?

**Review by bdhirsh — APPROVED** [review]
very cool!

**Review by bdhirsh — APPROVED** [review]
very cool!

File: torch/csrc/utils/python_raii.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+namespace impl {
+
+template <typename GuardT, typename... Args>
+struct RAIIContextManager {
+  explicit RAIIContextManager(Args&&... args)
+      : args_(std::forward<Args>(args)...) {}
+
+  void enter() {
+    auto emplace = [&](Args... args) {
+      return guard_.emplace(std::forward<Args>(args)...);
```
Comment (swolchok) [inline torch/csrc/utils/python_raii.h:15]: strangely, this doesn't seem to compile when c10::optional is std::optional (#101995). haven't figured out exactly why yet.

**Pull request closed without merging**